File: nm.py
from numpy import array, zeros, shape, ones, arange

# return codess for neldermead
SUCCESS = 0
BREAK   = 1
MAX_FUNC_EVALS = 2
MAX_ITER = 3

# Nelder-Mead algorithm constants
RHO   = 1
CHI   = 2
PSI   = 0.5
SIGMA = 0.5

def dumbreporter(what, itercount, func_evals, details, v, fv):
  print "=================================================="
  print "Iteration %d (%d func. evals): %s / %s\n" % (itercount, func_evals, what, details)
  print "SIMPLEX: ", v
  print "VALUES:  ", fv
  print "--------------------------------------------------"
  return False
  
# _change_simplex finds the direction where to go and transforms the simplex
#
# v    - an NxN+1-sized matrix with the simplex 
# fv   - an N+1-sized vector. fv[i] = func(v[:,i])
# func - cost function.

# returns a tuple ('transformation', number of function evaluations made)
def _change_simplex(v, fv, func):
  # handy func to replace the worst point with a new one
  def replace_worst(newx, newf, simplex=v, values=fv):
    simplex[:,-1] = newx
    values[-1] = newf

  evals = [0]

  # evaluate function and calculate the number of evaluations
  def evaluate(x, evals=evals):
    evals[0] += 1
    return func(x)

  n = shape(v)[0]
  
  # perform a shrink
  def shrink(simplex=v, values=fv, n=n):
    for j in range(1, n+1):
      simplex[:,j] = simplex[:,0] + SIGMA * (simplex[:,j] - simplex[:,0])
      values[j] = evaluate(simplex[:,j])

  # xbar = average of the n (NOT n+1) best points
  xbar = v[:, 0:n].mean(1)
  worst = v[:,-1]

  # Compute the reflection point
  xr = (1 + RHO)*xbar - RHO*worst
  fxr = evaluate(xr)

  # >>> Check the reflection point against our current best
  if fxr < fv[0]:
    # Calculate the expansion point
    xe = (1 + RHO*CHI)*xbar - RHO*CHI*worst
    fxe = evaluate(xe)

    if fxe < fxr:
      replace_worst(xe, fxe)
      how = 'expand'
    else:
      replace_worst(xr, fxr)
      how = 'reflect'

    return how, evals[0]

  # >>> Continuing, the reflection point is worse than the current best
  if fxr < fv[-2]:
    # >>> but at least it's better than the second worst point from the end
    replace_worst(xr, fxr)

    return 'reflect', evals[0]

  # >>> Worse than the second worse point from the end
  
  # Perform contraction
  if fxr < fv[-1]:
    # >>> better than the worst point we had

    # Perform an outside contraction
    xc = (1 + PSI*RHO)*xbar - PSI*RHO*worst
    fxc = evaluate(xc)
          
    if fxc <= fxr:
        replace_worst(xc, fxc)
        return 'contract outside', evals[0]
    else:
        # perform a shrink
        shrink()
        return 'shrink', evals[0]
    
  # Reflection bad, really bad, even worse than the worst one

  # Perform an inside contraction
  xcc = (1-PSI)*xbar + PSI*worst
  fxcc = evaluate(xcc)
          
  if fxcc < fv[-1]:
    replace_worst(xcc, fxcc)
    return 'contract inside', evals[0]

  shrink()
  return 'shrink', evals[0]

def neldermead(x0, func, outputfunc = None, maxiter = -200, maxfunevals = -200, \
               tolx = 1e-4, tolfun = 1e-4, funvalcheck = False):
  n = shape(x0)[0]

  if maxiter < 0:     maxiter     *= -n
  if maxfunevals < 0: maxfunevals *= -n

  # Set up a simplex near the initial guess.
  v = zeros((n, n+1))
  fv = zeros(n+1)
 
  # Place input guess in the simplex! (credit L.Pfeffer at Stanford)
  v[:,0] = x0
  fv[0] = func(x0)
  
  func_evals = 1
  itercount = 0

  # Initial simplex setup continues later

  # calls output and returns true if we need to break
  def call_output(place, details):
    return outputfunc and outputfunc(place, itercount, func_evals, details, v, fv) 

  if call_output('init', ''): return BREAK

  # 0th iteration, initial f(x)
  if call_output('iter', ''): return BREAK

  # Continue setting up the initial simplex.
  # Following improvement suggested by L.Pfeffer at Stanford
  
  usual_delta = 0.05              # 5 percent deltas for non-zero terms
  zero_term_delta = 0.00025       # Even smaller delta for zero elements of x

  for j in range(n):
    y = x0.copy()

    if y[j] != 0.0:
      y[j] *= (1 + usual_delta)
    else:
      y[j] = zero_term_delta

    v[:,j+1] = y
    fv[j+1] = func(y)

  # sort so v[0,:] has the lowest function value
  def sort_points(simplex, values):
    indexes = range(shape(simplex)[1])
    indexes.sort(key=lambda i: values[i])
    return simplex[:, indexes], values[indexes]

  v, fv = sort_points(v, fv)

  itercount += 1
  func_evals = n+1
  
  if call_output('iter', 'initial simplex'): return BREAK

  # Main algorithm
  # Iterate until the diameter of the simplex is less than tolx
  #   AND the function values differ from the min by less than tolf,
  #   or the max function evaluations are exceeded. (Cannot use OR instead of
  #   AND.)
  while func_evals < maxfunevals and itercount < maxiter:
    if abs(fv[0] - fv[1:n+1]).max() <= tolfun and abs(v[:,1:n+1] - v[:,0]).max() <= tolx:
      break

    what, evals = _change_simplex(v, fv, func)
    func_evals += evals

    v, fv = sort_points(v, fv)
    itercount += 1

    if call_output('iter', what): 
      return BREAK

  call_output('done', '')

  if func_evals >= maxfunevals:
    return MAX_FUNC_EVALS

  if itercount >= maxiter:
    return MAX_ITER

  return SUCCESS

File: nm.cpp
#include "nm.h"
#include <boost/numeric/ublas/matrix_proxy.hpp>
#include <boost/numeric/ublas/vector_proxy.hpp>
#include <vector>

using namespace boost::numeric::ublas;

namespace minimize { namespace nelder_mead {

// Nelder-Mead parameters
const double RHO   = 1;
const double CHI   = 2;
const double PSI   = 0.5;
const double SIGMA = 0.5;

// sort so simplex column 0 has the lowest function value
static void sort_simplex(values_vec &FV, simplex_mat &V);
static int change_simplex(values_vec &FV, simplex_mat &V, 
		cost_function func, void *farg, report_details &det);

struct BreakException {};

void run(const params_vector &x0, cost_function func, void *farg, 
		output &out, report_func reporter,
        int maxiter, int maxfunevals, double tolx, double tolfun)
{
    int n = x0.size();

    if (maxiter < 0)
        maxiter *= -n;

    if (maxfunevals < 0)
        maxfunevals *= -n;
  
    // Set up a simplex near the initial guess.
    simplex_mat V(n, n+1);
    values_vec FV(n+1);

    // Place input guess in the simplex! (credit L.Pfeffer at Stanford)
    column(V, 0) = x0;
    FV[0] = func(x0, farg);
  
    int func_evals = 1;
    int itercount = 0;

    // Initial simplex setup continues later
    try {
        if (reporter && 
                (reporter(INIT, NOTHING, itercount, func_evals, FV, V) || 
                 reporter(ITER, NOTHING, itercount, func_evals, FV, V)))
            throw BreakException();

        // Continue setting up the initial simplex.
        // Following improvement suggested by L.Pfeffer at Stanford

        const double USUAL_DELTA = 0.05;          // 5 percent deltas for non-zero terms
        const double ZERO_TERM_DELTA = 0.00025;   // Even smaller delta for zero elements of x

        for (int j = 0; j < n; ++j) {
            params_vector y = x0;

            if (y[j] != 0.0)
                y[j] *= 1.0 + USUAL_DELTA;
            else
                y[j] = ZERO_TERM_DELTA;

            column(V, j+1) = y;
            FV[j+1] = func(y, farg);
        }

        sort_simplex(FV, V);

        ++itercount;
        func_evals = n+1;

        if (reporter && reporter(ITER, INITIAL_SIMPLEX, itercount, func_evals, FV, V))
            throw BreakException();
  
        // Main algorithm
        // Iterate until the diameter of the simplex is less than tolx AND the
        // function values differ from the min by less than tolf, or the max
        // function evaluations are exceeded. (Cannot use OR instead of AND.)
        while (func_evals < maxfunevals && itercount < maxiter) {
            double FVdiff = norm_inf(subrange(FV, 1, n) - scalar_vector<double>(n-1, FV[0]));
            
            simplex_mat V2 = subrange(V, 0,n, 1,n+1);

            for (int j = 0; j < n; ++j)
                column(V2, j) -= column(V, 0);

            double Vdiff = norm_inf(V2);

            if (FVdiff <= tolfun && Vdiff <= tolx)
                break;

            report_details det;
            func_evals += change_simplex(FV, V, func, farg, det);

            sort_simplex(FV, V);

            ++itercount;

            if (reporter && reporter(ITER, det, itercount, func_evals, FV, V))
                throw BreakException();
        }

        if (reporter && reporter(DONE, NOTHING, itercount, func_evals, FV, V)) 
            ;

        if (func_evals >= maxfunevals)
            out.code = MAX_FUNC_EVALS;
        else if (itercount >= maxiter)
            out.code = MAX_ITER;
        else
            out.code = SUCCESS;
    }

    catch (BreakException) {
        out.code = ABORT;
    }

    out.last_x           = column(V, 0);
    out.last_f           = FV[0];
    out.iterations       = itercount;
    out.func_evaluations = func_evals;
}

class SortByCostFunctionValue {
public:
	SortByCostFunctionValue(const values_vec &_FV) : FV(_FV) {}

	bool operator()(int a, int b) {
		return FV[a] < FV[b];
	}
private:
	const values_vec &FV;
};

// sort so simplex column 0 has the lowest function value
static void sort_simplex(values_vec &FV, simplex_mat &V)
{
	values_vec FVcopy = FV;
	simplex_mat Vcopy = V;
	std::vector<int> indexes(FV.size());
	int j;

	for (j = 0; j < indexes.size(); ++j)
		indexes[j] = j;

	sort(indexes.begin(), indexes.end(), SortByCostFunctionValue(FV));

	for (j = 0; j < indexes.size(); ++j) {
		FV[j] = FVcopy[indexes[j]];
		column(V, j) = column(Vcopy, indexes[j]);
	}
}

#define REPLACE_AND_RETURN(newx, newf, detail) \
	do { \
		column(V, n) = newx; \
		FV[n] = newf; \
		det = detail; \
		return func_evals; \
	} while(0)

#define SHRINK_AND_RETURN(blah) \
	do { \
		func_evals += perform_shrink(FV, V, func, farg); \
		det = SHRINK; \
		return func_evals; \
	} while(0)

static int perform_shrink(values_vec &FV, simplex_mat &V, cost_function func, void *farg);

static int change_simplex(values_vec &FV, simplex_mat &V, 
		cost_function func, void *farg, report_details &det)
{
    int func_evals = 0;
	int j, n = FV.size()-1;
	
  	// xbar = average of the n (NOT n+1) best points
	params_vector xbar = zero_vector<double>(n), worst = column(V, n);

	for (j = 0; j < n; ++j)
		xbar += column(V, j);

	xbar /= double(n);
 
	// Compute the reflection point
 	params_vector xr = (1 + RHO)*xbar - RHO*worst;
	double fxr = func(xr, farg); ++func_evals;
	
  	/***** Check the reflection point against our current best *****/
  	if (fxr < FV[0]) {
    	// Calculate the expansion point
    	params_vector xe = (1.0 + RHO*CHI)*xbar - RHO*CHI*worst;
	    double fxe = func(xe, farg); ++func_evals;

	    if (fxe < fxr)
			REPLACE_AND_RETURN(xe, fxe, EXPAND);
		else
			REPLACE_AND_RETURN(xr, fxr, REFLECT);
	}
  
	/***** Continuing, the reflection point is worse than the current best *****/
	if (fxr < FV[n-1])
	    /* ... but at least it's better than the second worst point from the end */
		REPLACE_AND_RETURN(xr, fxr, REFLECT);
  
	/***** Worse than the second worse point from the end *****/

  	// Perform contraction
	if (fxr < FV[n]) {
		/***** better than the worst point we had *****/
    	
		// Perform an outside contraction
		params_vector xc = (1 + PSI*RHO)*xbar - PSI*RHO*worst;
		double fxc = func(xc, farg); ++func_evals;
          
		if (fxc <= fxr)
			REPLACE_AND_RETURN(xc, fxc, CONTRACT_OUTSIDE);
		else
			SHRINK_AND_RETURN();
	}

	/***** Reflection bad, really bad, even worse than the worst one *****/

	// Perform an inside contraction
	params_vector xcc = (1-PSI)*xbar + PSI*worst;
	double fxcc = func(xcc, farg); ++func_evals;

	if (fxcc < FV[n])
		REPLACE_AND_RETURN(xcc, fxcc, CONTRACT_INSIDE);

	SHRINK_AND_RETURN();
}

static int perform_shrink(values_vec &FV, simplex_mat &V, cost_function func, void *farg)
{
	params_vector best = column(V, 0);

	for (int j = 1; j < FV.size(); ++j) {
		column(V, j) = best + SIGMA * (column(V, j) - best);
		FV[j] = func(column(V, j), farg);
	}
	
	// number of function evaluations
	return FV.size() - 1;
}

} }
